pkl-core/src/main/java/org/pkl/core/http/NoProxyRule.java

```diff
@@ -0,0 +1,177 @@
+  private static final Pattern ipv6AddressOrCidr =
+      Pattern.compile("^(" + ipv6AddressString + ")(?:/([1-9][0-9]{0,2}))?$");
+
+  private @Nullable Integer ipv4 = null;
+  private @Nullable Integer ipv4Mask = null;
+  private @Nullable BigInteger ipv6 = null;
+  private @Nullable BigInteger ipv6Mask = null;
+  private @Nullable String hostname = null;
+  private boolean allNoProxy = false;
+
+  public NoProxyRule(String repr) {
+    if (repr.equals("*")) {
+      allNoProxy = true;
+      return;
+    }
+    var ipv4Matcher = ipv4AddressOrCidr.matcher(repr);
+    if (ipv4Matcher.matches()) {
+      var ipAddress = ipv4Matcher.group(1);
+      ipv4 = parseIpv4(ipAddress);
+      if (ipv4Matcher.groupCount() == 2 && ipv4Matcher.group(2) != null) {
+        var prefixLength = Integer.parseInt(ipv4Matcher.group(2));
+        if (prefixLength > 32) {
+          // best-effort (don't fail on invalid cidrs).
+          hostname = repr;
+        }
+        ipv4Mask = 0xffffffff << (32 - prefixLength);
+      }
+      return;
+    }
+    var ipv6Matcher = ipv6AddressOrCidr.matcher(repr);
+    if (ipv6Matcher.matches()) {
+      var ipAddress = ipv6Matcher.group(1);
+      ipv6 = parseIpv6(ipAddress);
+      if (ipv6Matcher.groupCount() == 4 && ipv6Matcher.group(4) != null) {
+        var maskBuffer = ByteBuffer.allocate(16).putLong(-1L).putLong(-1L);
+        var prefixLength = Integer.parseInt(ipv6Matcher.group(4));
+        if (prefixLength > 128) {
+          // best-effort (don't fail on invalid cidrs).
+          hostname = repr;
+          return;
+        }
+        ipv6Mask = new BigInteger(1, maskBuffer.array()).not().shiftRight(prefixLength);
+      }
+      return;
+    }
+    if (repr.startsWith(".")) {
+      hostname = repr.substring(1);
+    } else {
+      hostname = repr;
+    }
+  }
+
+  /** Tells if the provided URI should not be proxied according to the rules described. */
+  public boolean matches(URI uri) {
+    if (allNoProxy) {
+      return true;
+    }
+    var host = uri.getHost();
+    if (host == null) {
+      return false;
+    }
+    if (host.equalsIgnoreCase(hostname)) {
+      return true;
+    }
+    if (hostname != null && endsWithIgnoreCase(host, "." + hostname)) {
+      return true;
+    }
+    return ipV6Matches(uri.getHost()) || ipV4Matches(uri.getHost());
+  }
+
+  private boolean endsWithIgnoreCase(String str, String suffix) {
+    var len = suffix.length();
+    return str.regionMatches(true, str.length() - len, suffix, 0, len);
+  }
+
+  private boolean ipV4Matches(String hostname) {
+    if (ipv4 == null) {
+      return false;
+    }
+    if (!ipv4Address.matcher(hostname).matches()) {
+      return false;
+    }
+    var address = parseIpv4(hostname);
+    if (ipv4 == address) {
```

Comment:
Any reason to use `==` here and not `equals`? `==` only works for small integers.